core/src/main/java/org/apache/iceberg/SchemaParser.java

```diff
@@ -134,6 +143,10 @@ static void toJson(Type type, JsonGenerator generator) throws IOException {
     }
   }
 
+  public static void toJsonWithId(Schema schema, JsonGenerator generator) throws IOException {
```

Comment:
This is only called in `TableMetadataParser` and I don't see why there should be two `toJson` methods. It isn't a problem for the v1 metadata to be written with the schema ID because extra fields are ignored. So I think that this method isn't needed and `toJson` should always pass the ID to the struct writer.